### PR TITLE
Use TYPE instead of RESOURCE in help string

### PR DESCRIPTION
**Pull request by hurf**
Fixes: #10630
For get and describe command in kubectl, use KIND in help string.

### CODE REVIEW

**Comment by erictune** [comment]
LGTM

**Comment by erictune** [comment]
ok to test

**Comment by derekwaynecarr** [comment]
I don't get this change.

You call "kubectl get pods foo"

The RESOURCE is "pods".  The returned KIND is "Pod".  When you create a
pod, you get back "RESOURCE/NAME" in CLI response.  For example,
"pods/foo".  When we register a storage object with the rest server, we map
a RESOURCE string not a KIND to a URL endpoint.  In the future, we could
have multiple RESOURCE endpoints return the same KIND of object so it will
be even more confusing if we make this change.

@deads2k @smarterclayton let me know if i am missing something here,
wouldn't be the first time, but this seems incorrect.

**Comment by hurf** [comment]
Something like 'kubectl get pods/foo' looks stranges, use plural to request  and get one member, it's not natural. I think the reason ctl accept plurals is for user's ease, not to restrict them. Accept that we have obvious distinction between resources and kind internally. But for MM interface, would a way following natural language be better?

**Comment by derekwaynecarr** [comment]
If your change did what you suggested I would be less inclined to argue, but it just changes help text by making the help text incorrect with the current structure of kubectl which is a relatively thin wrapper over the Rest API at this time.  We could argue that it should take KIND but the fact remains that it actually aligns with RESOURCE for v1.

**Comment by erictune** [comment]
@derekwaynecarr  
Okay, I understand why you don't want to call this `KIND`.  And I don't mind if we drop this change.  But, I don't think "pods" in this context is a resource either.  Here are two definitions:
 - docs/api-conventions.md defines resource as "a representation of a system entity, sent or retrieved as JSON via HTTP to the server."  
 - [rfc 3986](https://tools.ietf.org/html/rfc3986) says a resource is "anything that might be identified by a URI". 
`pods` as a word by itself does not meet either definition.

**Comment by smarterclayton** [comment]
To be even more precise, we're talking about the resource collection name
(pods).  We definitely have a 1-1 mapping between kinds and resource
collection names (for all primary resource types), and a 1-1 primary
mapping for most resource collection names to kinds.  But there can and
will be resource collection names that don't correspond to Kinds and vice
versa.

You *can* feed a kind to the generic commands, but you can also give it a
resource collection name which does not match the kind it returns (which is
a perfectly reasonable restful thing to do).

One benefit of KIND is that it's much shorter... There may be an argument
on that alone for being a little imprecise.

On Jul 9, 2015, at 10:42 PM, Eric Tune <notifications@github.com> wrote:

@derekwaynecarr <https://github.com/derekwaynecarr>

Okay, I understand why you don't want to call this KIND. And I don't mind
if we drop this change. But, I don't think "pods" in this context is a
resource either. Here are two definitions:

   - docs/api-conventions.md defines resource as "a representation of a
   system entity, sent or retrieved as JSON via HTTP to the server."
    - rfc 3986 <https://tools.ietf.org/html/rfc3986> says a resource is
   "anything that might be identified by a URI". pods as a word by itself
   does not meet either definition.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10742#issuecomment-120207056>
.

**Comment by deads2k** [comment]
> Something like 'kubectl get pods/foo' looks stranges, use plural to request and get one member

Singular values are accepted (`kubectl get service/foo`) and https://github.com/GoogleCloudPlatform/kubernetes/pull/10652 is adding a utility method to convert to singular resource names, so output can be cleaned up as well.

**Comment by erictune** [comment]
please rebase

**Comment by bgrant0607** [comment]
Do we have a consensus that this a good or at least reasonable change to make?

If so, please rebase and look carefully for all occurrences:
```
$ grep -r RESOURCE pkg/kubectl
pkg/kubectl/cmd/scale.go:               Use: "scale [--resource-version=version] [--current-replicas=count] --replicas=COUNT RESOURCE NAME",
pkg/kubectl/cmd/scale.go:               return cmdutil.UsageError(cmd, "--replicas=COUNT RESOURCE NAME")
pkg/kubectl/cmd/expose.go:              Use:     "expose RESOURCE NAME --port=port [--protocol=TCP|UDP] [--target-port=number-or-name] [--name=name] [--public-ip=ip] [--type=type]",
pkg/kubectl/cmd/label.go:               Use:     "label [--overwrite] RESOURCE NAME KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]",
pkg/kubectl/cmd/describe.go:$ kubectl describe RESOURCE NAME_PREFIX
pkg/kubectl/cmd/describe.go:will first check for an exact match on RESOURCE and NAME_PREFIX. If no such resource
pkg/kubectl/cmd/describe.go:            Use:     "describe (RESOURCE NAME_PREFIX | RESOURCE/NAME)",
pkg/kubectl/cmd/stop.go:                Use:     "stop (-f FILENAME | RESOURCE (NAME | -l label | --all))",
pkg/kubectl/cmd/patch.go:               Use:     "patch RESOURCE NAME -p PATCH",
pkg/kubectl/cmd/delete.go:              Use:     "delete ([-f FILENAME] | (RESOURCE [(NAME | -l label | --all)]",
pkg/kubectl/cmd/get.go:         Use:     "get [(-o|--output=)json|yaml|template|wide|...] (RESOURCE [NAME] | RESOURCE/NAME ...)",
```

We should also document the resource/name syntax more consistently.

**Comment by smarterclayton** [comment]
I'm ok with KIND (I reserve the right to regret my decision later)

On Mon, Jul 27, 2015 at 1:51 AM, Brian Grant <notifications@github.com>
wrote:

> Do we have a consensus that this a good or at least reasonable change to
> make?
>
> If so, please rebase and look carefully for all occurrences:
>
> $ grep -r RESOURCE pkg/kubectl
> pkg/kubectl/cmd/scale.go:               Use: "scale [--resource-version=version] [--current-replicas=count] --replicas=COUNT RESOURCE NAME",
> pkg/kubectl/cmd/scale.go:               return cmdutil.UsageError(cmd, "--replicas=COUNT RESOURCE NAME")
> pkg/kubectl/cmd/expose.go:              Use:     "expose RESOURCE NAME --port=port [--protocol=TCP|UDP] [--target-port=number-or-name] [--name=name] [--public-ip=ip] [--type=type]",
> pkg/kubectl/cmd/label.go:               Use:     "label [--overwrite] RESOURCE NAME KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]",
> pkg/kubectl/cmd/describe.go:$ kubectl describe RESOURCE NAME_PREFIX
> pkg/kubectl/cmd/describe.go:will first check for an exact match on RESOURCE and NAME_PREFIX. If no such resource
> pkg/kubectl/cmd/describe.go:            Use:     "describe (RESOURCE NAME_PREFIX | RESOURCE/NAME)",
> pkg/kubectl/cmd/stop.go:                Use:     "stop (-f FILENAME | RESOURCE (NAME | -l label | --all))",
> pkg/kubectl/cmd/patch.go:               Use:     "patch RESOURCE NAME -p PATCH",
> pkg/kubectl/cmd/delete.go:              Use:     "delete ([-f FILENAME] | (RESOURCE [(NAME | -l label | --all)]",
> pkg/kubectl/cmd/get.go:         Use:     "get [(-o|--output=)json|yaml|template|wide|...] (RESOURCE [NAME] | RESOURCE/NAME ...)",
>
> We should also document the resource/name syntax more consistently.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10742#issuecomment-125094571>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by derekwaynecarr** [comment]
Can it be called something other than KIND or RESOURCE?  It seems clear to me that it is neither.

**Comment by bgrant0607** [comment]
Type or object: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/api-conventions.md#types-kinds

Or we could introduce "collection".

But I honestly wouldn't be too pedantic for a CLI where most users won't read and internalize the API specification.

**Comment by erictune** [comment]
Any objections to `TYPE`?

**Comment by hurf** [comment]
I rebased the change with KIND. If there're other ideas, I'm happy to modify it again. 
Personally instead of `KIND` or 'TYPE' I prefer `RESOURCE_TYPE`, from a user's perspective I feel it's more clear.

**Comment by derekwaynecarr** [comment]
+1 ```TYPE```

**Comment by erictune** [comment]
Fine with `RESOURCE_TYPE` too.

**Comment by bgrant0607** [comment]
RESOURCE_TYPE is clear, but kind of wide.

**Comment by smarterclayton** [comment]
In the CLI context TYPE or KIND seem ok - RESOURCE_TYPE is too long and
also still vague (what's a resource).  TYPE is more generic than KIND, and
would be more familiar to most users.

On Wed, Jul 29, 2015 at 4:01 PM, Brian Grant <notifications@github.com>
wrote:

> RESOURCE_TYPE is clear, but kind of wide.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10742#issuecomment-126080067>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by hurf** [comment]
Since most people are positive with `TYPE`, changed with `TYPE`.

File: pkg/kubectl/cmd/get.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ import (
 const (
 	get_long = `Display one or many resources.
 
-Possible resources include (case insensitive): pods (po), services (svc),
+Possible resource kinds include (case insensitive): pods (po), services (svc),
```
Comment (erictune) [inline pkg/kubectl/cmd/get.go]: s/resource kinds/types/

**Comment by erictune** [comment]
One change and then regen docs and then it will look good to me.

**Comment by erictune** [comment]
lgtm, thanks for the fix.

**Pull request merged**